Add QR-DQN

ku2482 opened the pull request:
<!--- Provide a general summary of your changes in the Title above (e.g. feature name, fix of a bug)-->
Implement QR-DQN

## Description
<!--- Describe your changes in detail, including links to any sources for new features -->
Paper: https://arxiv.org/abs/1710.10044
- [x] Implement QR-DQN
- [x] Add quantile_huber_loss to `sb3_contrib.common.utils`
- [x] Add document for QR-DQN
- [x] Add benchmark

## Context
<!--- Link the related issue here. You can use the syntax `closes #100` if this solves the issue #100 -->
- [x] I have raised an issue to propose this change ([required](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/CONTRIBUTING.md))

closes #12 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've read the [CONTRIBUTION](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] The functionality/performance matches that of the source (**required** for new training algorithms or training-related features).
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have included an example of using the feature (*required for new features*).
- [ ] I have included baseline results (**required** for new training algorithms or training-related features).
- [ ] I have updated the documentation accordingly.
- [x] I have updated the changelog accordingly (**required**).
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)


Note: we are using a maximum length of 127 characters per line

### Benchmark


|Environments|   QRDQN   | DQN
|------------|-----------| -----------|
|            |logs_qrdqn/| |
|Breakout    |413 +/- 21 | ~300 |
|Pong        |20 +/- 0   | ~20 |

|Environments|    DQN    |   QRDQN   |
|------------|-----------|-----------|
|            |logs_qrdqn/|logs_qrdqn/|
|CartPole    |386 +/- 64 |500 +/- 0  |
|MountainCar |-111 +/- 4 |-107 +/- 4 |
|LunarLander |168 +/- 39| 195 +/- 28|
|Acrobot     |-73 +/- 2  |-74 +/- 2  |




![Results_Breakout](https://user-images.githubusercontent.com/1973948/102275723-22cb7b80-3f26-11eb-8384-10b74750dfce.png)
![Results_Pong](https://user-images.githubusercontent.com/1973948/102275732-252dd580-3f26-11eb-987b-57b0553c13bd.png)

![Results_Acrobot](https://user-images.githubusercontent.com/1973948/102718704-cbdcf200-42e9-11eb-8130-64f5e6a8bbfc.png)
![Results_CartPole](https://user-images.githubusercontent.com/1973948/102718707-cc758880-42e9-11eb-8075-7200ed42a383.png)
![Results_LunarLander](https://user-images.githubusercontent.com/1973948/102718708-cd0e1f00-42e9-11eb-932e-bd732dd106e1.png)
![Results_MountainCar](https://user-images.githubusercontent.com/1973948/102718709-cd0e1f00-42e9-11eb-8cc8-3843e6b1fa0c.png)



<!--- This Template is an edited version of the one from https://github.com/evilsocket/pwnagotchi/ -->

araffin commented on sb3_contrib/__init__.py:
```diff
@@ -1,6 +1,7 @@
 import os
 
 # from sb3_contrib.cmaes import CMAES
```
minor: could you remove that commented line? (my fault)

araffin commented on sb3_contrib/qrdqn/policies.py:
```diff
@@ -0,0 +1,244 @@
+        features_extractor_kwargs: Optional[Dict[str, Any]] = None,
+        normalize_images: bool = True,
+        optimizer_class: Type[th.optim.Optimizer] = th.optim.Adam,
+        optimizer_kwargs: Dict[str, Any] = {"eps": 0.01 / 32},
```
Please avoid mutable as default argument
why `0.01 / 32`, it depends on the minibatch size?

See https://github.com/satwikkansal/wtfpython#-beware-of-default-mutable-arguments

ku2482 replied:
>why 0.01 / 32, it depends on the minibatch size?

It's proposed in the paper (they swept it in  (0.01/32, 0.005/32, 0.001/32)), and 32 is the minibatch size.

araffin replied:
ok, probably put a comment too then ;)

ku2482 replied:
What is the best way to implement default "eps"?

araffin replied:
>What is the best way to implement default "eps"?


there are different ways...
we do it like that in A2C: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/a2c/a2c.py#L111

ku2482 replied:
Thank you :)

araffin commented on sb3_contrib/common/utils.py:
```diff
@@ -0,0 +1,30 @@
+    n_quantiles = current_quantile.shape[-1]
+    tau = (th.arange(n_quantiles, device=current_quantile.device).float() + 0.5) / n_quantiles
+
+    if current_quantile.ndim == 3:
```
TQC has more dimensions because there are several critics?

ku2482 replied:
Yes.

araffin replied:
then probably add a small comment next to the `For TQC`

araffin reviewed (commented):
Overall, look good already =)

Obviously missing tests, documentation and benchmark but it is a good start!

Did you had any issue so far? or things that slow you down because of how SB3 work in implementing QR-DQN?
(asking for what we can improve in SB3 ;))

ku2482 commented:
@araffin 
Hi, I have implemented QR-DQN roughly, and I have some points I want to discuss.

1
I named some classes and variables like `QuantileNetwork` and `current_quantile`, however in TQC there are some variables like `next_z`. Could you tell me what names do you prefer?

2
I added `quantile_huber_loss` to `sb3_contrib.common.utils`. However, if you plan to add IQN([paper](https://arxiv.org/abs/1806.06923)) later, `tau` should be an argument. What do you think?

3
I didn't inherit any classes from DQN's for simplicity. I'm familiar with RL, but not with software architecture, and I'm not sure if I should inherit and re-use it. Could you give me any advice?

4
I don't have enough computational resources to benchmark on Atari.
Could you help me test the performance once I finish implementing it?

(I only have 16GB of RAM and I usually store images as a list of LazyFrames, which uses 4 times less memory. Maybe it's better to add such a feature to reduce memory usage?)

Thanks ;)

araffin commented:
>, however in TQC there are some variables like next_z. Could you tell me what names do you prefer?

`current_quantile` is must better ;) (please change TQC too so the two matches, probably `next_quantiles` if it is `next_z`)

> I added quantile_huber_loss to sb3_contrib.common.utils. However, if you plan to add IQN(paper) later, tau should be an argument. What do you think?

what is tau exactly? (it would be nice if we have a better name, I'm quite new to quantile regression, so everything is not too clear yet)
yes, it sounds good if it does not add too much complexity.
(I need to re-read IQN later)

>I didn't inherit any classes from DQN's for simplicity.

that's fine ;)
It is part of the design choices we made for SB3.
We favor simplicity and readability over modularity. (even though we try to avoid code duplication when possible)

> Could you help me test the performance once I finish implementing it?

yes ;) but please test it first on simpler env and with a smaller replay buffer size.

>I only have 16GB of RAM and I usually store images as a list of LazyFrames, which uses 4 times less memory. Maybe it's better to add such a feature to reduce memory usage?

Look at the zoo and the replay buffer, we have a `optimize_memory_usage` argument for that ;)

ku2482 commented:
>please change TQC too so the two matches, probably next_quantiles if it is next_z

>We favor simplicity and readability over modularity. (even though we try to avoid code duplication when possible)

> yes ;) but please test it first on simpler env and with a smaller replay buffer size.

Sure, thank you for your advice.

>what is tau exactly?

We model the quantile function, which is the mapping from the cumulative probability to the quantile value.
`tau` is the cumulative probabilities, which are fixed at equal intervals in TQC and QR-DQN (not in IQN).
So, `cum_p` or `cum_prob` would be a good candidate?

>Look at the zoo and the replay buffer, we have a optimize_memory_usage argument for that ;)

I didn't notice it. It seems it stores the next observation and observation as the same array, and so efficient 👍 

>  I'm not sure for the lazy frame, but if it is simple enough to implement,

[LazyFrames](https://github.com/Stable-Baselines-Team/stable-baselines/blob/master/stable_baselines/common/atari_wrappers.py#L244) is a list of arrays, and convert a list into a (frame-stacked) array when called. Because we store it as a list of arrays, which exactly is a list of references to arrays, it never store the same frame twice.
It's really simple to implement as [this](https://github.com/ku2482/sac-discrete.pytorch/blob/master/sacd/memory/base.py#L46), but it is a little bit slow.


Thank you for your kind response.

araffin commented:
> tau is the cumulative probabilities, which are fixed at equal intervals in TQC and QR-DQN (not in IQN).

thanks for the refresher =).
So, if I remember correctly, this is the case for QR-DQN because each quantile is a "unit quantile" (in the sense each of them have the same weights)?

>cum_prob would be a good candidate?

yep + a comment ;) (in case of doubt, more verbose name is always good ;))

>I didn't notice it. It seems it stores the next observation and observation as the same array, and so efficient +1

yes, it is not as memory efficient as LazyFrame but it is as fast as a normal buffer and works also without frame stacking at the cost of some complexity (that's why it is False by default).

>It's really simple to implement as this, but it is a little bit slow.

how much slower? (1.2x or ~2x slower)

ku2482 commented:
> So, if I remember correctly, this is the case for QR-DQN because each quantile is a "unit quantile" (in the sense each of them have the same weights)?

Yes 👍 

> how much slower? (1.2x or ~2x slower)

I will test it again in a couple of days.

araffin commented on sb3_contrib/qrdqn/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from sb3_contrib.qrdqn.policies import CnnPolicy, MlpPolicy
```
you will need to update `setup.cfg` to ignore the warnings of the linter ;)

araffin commented on sb3_contrib/qrdqn/policies.py:
```diff
@@ -0,0 +1,248 @@
+            "normalize_images": normalize_images,
+        }
+
+        if optimizer_class is th.optim.Adam and "eps" not in self.optimizer_kwargs:
```
I would rather do it in the contructor of `QRDQN` so you also have access to the batch size there ;)

araffin commented on sb3_contrib/qrdqn/qrdqn.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    Quantile Regression Deep Q-Network (QR-DQN)
+    Paper: https://arxiv.org/abs/1710.10044
+    Default hyperparameters are taken from the paper
```
```suggestion
    Default hyperparameters are taken from the paper
    
```

araffin replied:
this is for the automatic documentation

araffin commented on sb3_contrib/qrdqn/qrdqn.py:
```diff
@@ -0,0 +1,244 @@
+                # Follow greedy policy: use the one with the highest value
+                target_quantile, _ = target_quantile.max(dim=2)
+                # 1-step TD target
+                target_quantile = replay_data.rewards + (1 - replay_data.dones) * self.gamma * target_quantile
```
probably rename it `quantile_backup` so you don't overwrite `target_quantile` defined above?
btw, why `target_quantile` (singular) and not `target_quantiles` ?

araffin commented on sb3_contrib/tqc/tqc.py:
```diff
@@ -242,19 +223,18 @@ def train(self, gradient_steps: int, batch_size: int = 64) -> None:
-                next_z = self.critic_target(replay_data.next_observations, next_actions)
-                sorted_z, _ = th.sort(next_z.reshape(batch_size, -1))
-                sorted_z_part = sorted_z[:, : self.critic.quantiles_total - top_quantiles_to_drop]
+                next_quantile = self.critic_target(replay_data.next_observations, next_actions)
```
dito (about singular/plural)

ku2482 replied:
>probably rename it quantile_backup so you don't overwrite target_quantile defined above?

(Belleman) "buckup" and "target" mean the same thing here, don't they?
For clarity, it may be good to name them `next_*` and `target_*`? (it would be incompatible with stable_baselines3)
What do you think?
```
next_* = ...  # Estimate values for next observation
target_* = ... next_* ...   # Something like "R + gamma * next_*", which is the target for the loss
```

>btw, why target_quantile (singular) and not target_quantiles ?

I was confused that some variables like `next_log_prob`, `current_q`, and `target_q` are singular in DQN and SAC, but others are plural. There was no particular intention, so I'd be happy to rename variables as you prefer :)

araffin replied:
>(Belleman) "buckup" and "target" mean the same thing here, don't they?

yes, I was using "target" in SAC/TD3 because the next q-value came from the target network.
but I would agree with renaming `target_` to `next_` (and maybe rename the first `next_target_quantiles`), what do you think?

(we would need to do that in https://github.com/DLR-RM/stable-baselines3/pull/245 to keep SB3 in sync)

>current_q, and target_q are singular in DQN and SAC, 

sorry for that, I guess the true name should be `current_q_estimates` (plural), maybe a thing to change in https://github.com/DLR-RM/stable-baselines3/pull/245 (by @erniejunior )
and `log_prob` is short for "log probabilities" (I truncated it to avoid too long names and equations on several lines) but I agree it is a bit confusing.

ku2482 replied:
> I was using "target" in SAC/TD3 because the next q-value came from the target network.
> but I would agree with renaming target_ to next_ (and maybe rename the first next_target_quantiles), what do you think?

I see, now it makes sense. But I assume people call it "target" because it's the target value for training, not because it comes from the target network.
In my opinion, because "buckup" is a less common word, it's clearer for me either
1. to name both "Q(s_t+1)" and "R + gamma * (max)Q(s_t+1)"  as `target_*`, as in [DQN's code](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/dqn/dqn.py#L157)
2. to name "Q(s_t+1)" as `next_*` and "R + gamma * (max)Q(s_t+1)" as `target_*`, as in (my) [TQC's code](https://github.com/ku2482/stable-baselines3-contrib/blob/feat/qrdqn/sb3_contrib/tqc/tqc.py#L220).

> sorry for that, I guess the true name should be current_q_estimates (plural)

I see. So it's natural to name `current_quantile` or `current_quantile_estimates`?

araffin replied:
I would go for 2. (and we will need to change that in https://github.com/DLR-RM/stable-baselines3/pull/245.)

>I see. So it's natural to name current_quantile or current_quantile_estimates?

here it is slightly different as you have several quantiles (the complete name of `current_q_estimates` would be `current_q_values_estimates` maybe `current_q_values` would be better in fact)

ku2482 replied:
>here it is slightly different as you have several quantiles (the complete name of current_q_estimates would be current_q_values_estimates maybe current_q_values would be better in fact)

I see. I'm not fluent in English, so I'll follow your preference.

ku2482 replied:
I pushed some commits to rename them, what do you think?
(I modified TQC's code somewhat to make it easy to understand, sorry about that...)

Also, how should I evaluate QR-DQN?
The evaluation protocol of sb3 seems different from the paper.
(The paper reported the best agent's score within 125k steps(500k frames))

Anyway, I really appreciate your kind support ;)

araffin commented on sb3_contrib/common/utils.py:
```diff
@@ -0,0 +1,36 @@
+    # Cumulative probabilities to calculate quantile values.
+    cum_prob = (th.arange(n_quantiles, device=current_quantiles.device, dtype=th.float) + 0.5) / n_quantiles
+    if current_quantiles.ndim == 3:
+        cum_prob = cum_prob.view(1, 1, -1, 1)  # For TQC
```
It would be nice to have the meaning of each dimension in comment (as it was done before)
and add the comment on why the shape for TQC is different that for QR-DQN

araffin commented on sb3_contrib/qrdqn/qrdqn.py:
```diff
@@ -0,0 +1,250 @@
+            current_quantiles = th.gather(current_quantiles, dim=2, index=actions).squeeze(2)
+
+            # Compute Quantile Huber loss
+            loss = quantile_huber_loss(current_quantiles, target_quantiles) * self.n_quantiles
```
why do you multiply by `self.n_quantiles`? (that was not done in TQC original code if I recall... I did not check for QR-DQN yet)

ku2482 replied:
We sum over a quantile dimension, which is common in QR-DQN, IQN, and FQF.
I'm not sure why they take the mean in TQC.

araffin replied:
It seems that Dopamine is using the mean loss by default: https://github.com/google/dopamine/blob/master/dopamine/jax/agents/quantile/quantile_agent.py#L91
The same goes for IQN implementation of Pfrl: https://github.com/pfnet/pfrl/blob/master/pfrl/agents/iqn.py#L211

or the Facebook ReAgent implementation: https://github.com/facebookresearch/ReAgent/blob/master/reagent/training/qrdqn_trainer.py#L157
maybe that could be a parameter? (and check how it affects the learning)

ku2482 replied:
> It seems that Dopamine is using the mean loss by default: https://github.com/google/dopamine/blob/master/dopamine/jax/agents/quantile/quantile_agent.py#L91
The same goes for IQN implementation of Pfrl: https://github.com/pfnet/pfrl/blob/master/pfrl/agents/iqn.py#L211

Their implementations sum over the quantile dimension, so the same as mine, aren't they?
Multiplying "n_quantiles" means that summing over the quantile dimension.

ku2482 replied:
Maybe we should add an argument like `sum_over_quantiles : bool` to quantile_huber_loss?

araffin replied:
>Their implementations sum over the (current) quantile dimension, so the same as mine, aren't they?

I will try to check later when I'm rested ;)

>Maybe we should add an argument like sum_over_quantiles : bool to quantile_huber_loss?


Probably, yes

araffin commented on sb3_contrib/qrdqn/qrdqn.py:
```diff
@@ -0,0 +1,250 @@
+            current_quantiles = self.quantile_net(replay_data.observations)
+
+            # Retrieve the quantiles for the actions from the replay buffer
+            actions = replay_data.actions[..., None].long().expand(batch_size, self.n_quantiles, 1)
```
please put a comment about what shape is actions after that transform

araffin commented on sb3_contrib/qrdqn/qrdqn.py:
```diff
@@ -0,0 +1,250 @@
+
+            # Retrieve the quantiles for the actions from the replay buffer
+            actions = replay_data.actions[..., None].long().expand(batch_size, self.n_quantiles, 1)
+            current_quantiles = th.gather(current_quantiles, dim=2, index=actions).squeeze(2)
```
```suggestion
            current_quantiles = th.gather(current_quantiles, dim=2, index=actions).squeeze(dim=2)
```

Please use explicit keywords when dealing with dimensions ;)

araffin commented on sb3_contrib/tqc/tqc.py:
```diff
@@ -237,24 +218,27 @@ def train(self, gradient_steps: int, batch_size: int = 64) -> None:
+                # td error + entropy term
+                target_quantiles = next_quantiles - ent_coef * next_log_prob.reshape(-1, 1)
+                target_quantiles = replay_data.rewards + (1 - replay_data.dones) * self.gamma * target_quantiles
+                # Make it briadcastable with current_quantiles
```
typo + please add a comment on what on the shape of target quantiles before and after the transform

araffin commented on sb3_contrib/tqc/tqc.py:
```diff
@@ -237,24 +218,27 @@ def train(self, gradient_steps: int, batch_size: int = 64) -> None:
+                target_quantiles = next_quantiles - ent_coef * next_log_prob.reshape(-1, 1)
+                target_quantiles = replay_data.rewards + (1 - replay_data.dones) * self.gamma * target_quantiles
+                # Make it briadcastable with current_quantiles
+                target_quantiles.unsqueeze_(1)
```
```suggestion
                target_quantiles.unsqueeze_(dim=1)
```

araffin commented on tests/test_cnn.py:
```diff
@@ -7,18 +7,18 @@
     env = FakeImageEnv(screen_height=40, screen_width=40, n_channels=1, discrete=model_class not in {TQC})
     kwargs = {}
-    if model_class in {TQC}:
+    if model_class in {TQC, QRDQN}:
```
we should probably also reduce the number of quantiles of QR-QDN during the tests, so they run faster, no?
(QR-DQN uses 200 quantiles by default vs 25 for TQC)

araffin commented on tests/test_cnn.py:
```diff
@@ -52,25 +59,27 @@ def params_should_differ(params, other_params):
 @pytest.mark.parametrize("model_class", [TQC])
```
not yet included?

araffin commented on tests/test_run.py:
```diff
@@ -38,3 +38,17 @@ def test_sde():
+    model = QRDQN(
+        "MlpPolicy",
+        "CartPole-v1",
+        policy_kwargs=dict(net_arch=[64, 64]),
```
dito for the number of quantiles ;)

araffin commented:
>I pushed some commits to rename them, what do you think?
(I modified TQC's code somewhat to make it easy to understand, sorry about that...)

Overall looks good ;) (I will check in details later)

>Also, how should I evaluate QR-DQN?
The evaluation protocol of sb3 seems different from the paper.
(The paper reported the best agent's score within 125k steps(500k frames))

True, the good news is that we can compare to [Intel Coach](https://github.com/IntelLabs/coach/tree/master/benchmarks/qr_dqn).

I would first compare on simple env (CartPole-v1, LunarLander-v2) and then go for Pong and then Breakout.
For the simpler env, you can probably take a look at [dopamine](https://github.com/google/dopamine) or Intel Coach.
Choose the one which is easier to setup and to tweak, for instance, I did [that](https://github.com/sfujim/TD3/compare/master...araffin:master) (integrating SB3 EvalCallback) to compare SB3 TD3 performance to original implementation.


>Anyway, I really appreciate your kind support ;)


your welcome =)

ku2482 commented:
I quickly tested on BreakoutNoFrameskip-v4 to compare with [coach](https://github.com/IntelLabs/coach/tree/master/benchmarks/qr_dqn). (I killed processes due to the limited resource.)
blue: `max_grad_norm=10.0(default)`, orange: `max_grad_norm: np.inf(paper)`

![スクリーンショット 2020-12-11 10 23 07](https://user-images.githubusercontent.com/37267851/101849417-fcf74d00-3b9a-11eb-99ea-6f7eea2326b6.png)

It seems fine compared to coach. So could you help me test QR-DQN in detail?
Here are the things I want to discuss.

- It's redundant if we clip the norm of the grad with `np.inf`. So maybe it's good if `max_grad_norm` can be disabled and we skip clipping? BTW, why DQN in sb3 uses `max_grad_norm=10.0` by default?

- Is it okay that the evaluation protocol for Atari is different from the paper?
I think (i) and (ii) can be applied to sb3.
  1. stop episodes at 27k steps
  2. use epsilon-greedy for evaluation (smaller epsilon than training, like 0.001)
  3. run for 125k steps and report the best episode score.

araffin commented:
>it's redundant if we clip the norm of the grad with np.inf. So maybe it's good if max_grad_norm can be disabled and we skip clipping? BTW, why DQN in sb3 uses max_grad_norm=10.0 by default?

yes, we can make it optional, especially if it is not mentioned in the paper.
The value comes from OpenAI Baselines implementation.

>Is it okay that the evaluation protocol for Atari is different from the paper?

yes, it should be at least the same as the one we used for checking DQN implementation:
so training for 1e7 steps (and periodic evaluation).

>stop episodes at 27k steps

why 27k steps?

>use epsilon-greedy for evaluation (smaller epsilon than training, like 0.001)

yes, we need to adapt the `EvalCallback` but I will do one run without first.


>EDIT: I used the code below.

Please create a branch in the rl-zoo next time, as mentioned in the contributing guide ;) (I think I will do it)

> Add the argument sum_over_quantiles(or other name)? Should I make cum_prob the optional argument for IQN?

`sum_over_quantiles` sounds good.
For the `cum_prob`, yes, make it probably an optional argument that you compute if not provided.

araffin commented on sb3_contrib/common/utils.py:
```diff
@@ -0,0 +1,40 @@
+    :return: the loss
+    """
+    if current_quantiles.ndim != target_quantiles.ndim:
+        raise ValueError(
```
minor: it would be also good to check inthe tests that `quantile_huber_loss` actually throw those errors when needed.
If necessary, I will do it ;)

ku2482 commented:
> why 27k steps?

[Double DQN paper](https://arxiv.org/abs/1509.06461) mentioned it as below, and most papers follow this evaluation protocol.

>We start an evaluation episode from each of these starting points and run the emulator for up to 108,000 frames (30 mins at 60Hz including the trajectory before the starting point).

> Please create a branch in the rl-zoo next time, as mentioned in the contributing guide ;) (I think I will do it)

Sorry about that, and thank you ;)

araffin commented:
The RL-Zoo branch is here ;)  https://github.com/DLR-RM/rl-baselines3-zoo/tree/feat/qr-dqn

araffin commented on docs/guide/algos.rst:
```diff
@@ -9,6 +9,7 @@ along with some useful characteristics: support for discrete/continuous actions,
 Name         ``Box``     ``Discrete`` ``MultiDiscrete`` ``MultiBinary`` Multi Processing
 ============ =========== ============ ================= =============== ================
 TQC          ✔️          ❌            ❌                ❌              ❌
+QRDQN        ️❌          ✔            ❌                ❌              ❌
```
pleas re-use the same character as above (copy-paste should work) and also update the README ;)

ku2482 replied:
You mean, use "QR-DQN", not "QRDQN"?

araffin replied:
yes and also the ✔️ seems to be different

araffin commented on docs/guide/examples.rst:
```diff
@@ -16,6 +16,22 @@ Train a Truncated Quantile Critics (TQC) agent on the Pendulum environment.
+
+  from sb3_contrib import QRDQN
+
+  env = gym.make("CartPole-v1")
+
+  model = QRDQN("MlpPolicy", "CartPole-v1", verbose=1)
```
```suggestion
  model = QRDQN("MlpPolicy", "CartPole-v1", verbose=1)
```

araffin commented on docs/modules/qrdqn.rst:
```diff
@@ -0,0 +1,102 @@
+  import numpy as np
+
+  from sb3_contrib import QRDQN
+  from sb3_contrib.qrdqn import MlpPolicy
```
Please use the string version (recommended as it is the same for each algorithm) and show how to modify the number of quantiles (cf TQC example in the doc)

araffin commented on docs/guide/examples.rst:
```diff
@@ -16,6 +16,22 @@ Train a Truncated Quantile Critics (TQC) agent on the Pendulum environment.
   model.learn(total_timesteps=10000, log_interval=4)
   model.save("tqc_pendulum")
 
+QR-DQN
```
Please also update the changelog (put your username next to each entry) and add your username at the bottom ;)

araffin reviewed (commented):
after a quick look, it seems we are only missing the benchmark on Atari and minor edits in the doc, and then we are good for merging =)

I also quickly tested QR-DQN in the zoo on classic control environment (with hyperparameters from DQN) and it works at least as well, so it's a good sign =)

ku2482 commented:
I really appreciate your help ;)

araffin commented on docs/modules/qrdqn.rst:
```diff
@@ -0,0 +1,101 @@
+  env = gym.make("CartPole-v1")
+
+  model = QRDQN("MlpPolicy", env, verbose=1)
+  model.learn(total_timesteps=10000, n_quantiles=50, log_interval=4)
```
thanks for editing, but this is not how it works, no? (again take a look at TQC example ;) )

ku2482 replied:
Sorry about that...

araffin commented:
My results on 3 seeds (Breakout and Pong) without smoothing window and with the default hyperparameters (so the exploration fraction is wrong and the epsilon for the evaluation is the same as for training):
![Results_Breakout](https://user-images.githubusercontent.com/1973948/102275723-22cb7b80-3f26-11eb-8384-10b74750dfce.png)
![Results_Pong](https://user-images.githubusercontent.com/1973948/102275732-252dd580-3f26-11eb-987b-57b0553c13bd.png)


Overall, it looks good as it matches Intel Coach results (+ a nice boost in final performance ~400 vs ~300 for DQN on Breakout) but as mentioned by @Miffyli that would nice to have at least one run with the original evaluation setting.

ku2482 commented:
Thank you for your support ;)

> but as mentioned by @Miffyli that would nice to have at least one run with the original evaluation setting.

Yeah, but it needs some changes to evaluation callbacks and takes some time.
Maybe, is it good to make another PR to add the evaluation callback for Atari, and add results of DQN, QR-DQN, and other discrete algorithms?

araffin commented:
> Maybe, is it good to make another PR to add the evaluation callback for Atari, and add results of DQN, QR-DQN, and other discrete algorithms?

yes, probably a good addition for another PR. But I would need @Miffyli advice on that if it is worth implementing ;) (as I'm mostly focused on Continuous Control and not really Atari games with discrete control...)

araffin commented:
I hope to have more time end of this week or next week to finish the review ;)

Anyway thank you already for your good work =)

araffin commented on sb3_contrib/common/utils.py:
```diff
@@ -0,0 +1,59 @@
+def quantile_huber_loss(
+    current_quantiles: th.Tensor,
+    target_quantiles: th.Tensor,
+    cum_prob: Optional[th.Tensor] = None,
```
those are also called "midpoints" in the paper, no?

ku2482 replied:
Yes in QR-DQN paper, but it's less common and not used in other papers (IQN, FQF).

araffin replied:
maybe just add a comment saying "also called midpoints in QR-DQN paper"

araffin commented on sb3_contrib/common/utils.py:
```diff
@@ -0,0 +1,59 @@
+    huber_loss = th.where(abs_pairwise_delta > 1, abs_pairwise_delta - 0.5, pairwise_delta ** 2 * 0.5)
+    loss = th.abs(cum_prob - (pairwise_delta.detach() < 0).float()) * huber_loss
+    if sum_over_quantiles:
+        loss = loss.sum(dim=-2).mean()
```
I think I know why TQC is using the mean directly, this allows to avoid large loss values (and also to have a loss magnitude that does not depends on the number of critics/quantiles),
which makes it probably easier to tune the learning rate.

ku2482 replied:
Probably yes ;)

araffin reviewed (approved):
I could finally do the review, LGTM, thanks =)
(only missing to add the benchmark to the doc)

ku2482 commented:
Thank you so much ;)

>only missing to add the benchmark to the doc

Could you do this for me, or should I update the doc?

araffin commented:
>Could you do this for me, or should I update the doc?

done ;)
I also added a small benchmark (5 seeds, tuned hyperparameters) on classic control tasks.

@Miffyli I let you decide if we need more experiments (even though we match Intel Coach results), or we merge ;)

Miffyli commented:
@araffin 

Visually comparing results here vs. original paper the Breakout and Pong results seem to match up till the trained 10e6, so I would trust this implementation enough to use it as QR-DQN myself and believe we can merge it :).

Just two things: docs should mention what we compared against (Intel Coach and, very roughly, original paper) and also update zoo-instructions once we merge the branch.

ku2482 commented:
@araffin @Miffyli 
Thank you for your kind support ;)

Squeemos commented:
Hate to post on very late, but is there a specific reason that we cannot use action_noise in QRDQN? Was looking to use it for exploration rather than a exploration scheduler, and would prefer using action noise over noisy nets for now.

Miffyli commented:
> Hate to post on very late, but is there a specific reason that we cannot use action_noise in QRDQN? Was looking to use it for exploration rather than a exploration scheduler, and would prefer using action noise over noisy nets for now.

QRDQN supports only Discrete action spaces, and there is no action_noise for Discrete spaces (at least no ones implemented in SB3 as of writing).

hh0rva1h commented:
> My results on 3 seeds (Breakout and Pong) without smoothing window and with the default hyperparameters (so the exploration fraction is wrong and the epsilon for the evaluation is the same as for training):
> Environments 	QRDQN
> 	logs_qrdqn/
> Breakout 	413 +/- 21
> Pong 	20 +/- 0
> 
> ![Results_Breakout](https://user-images.githubusercontent.com/1973948/102275723-22cb7b80-3f26-11eb-8384-10b74750dfce.png) ![Results_Pong](https://user-images.githubusercontent.com/1973948/102275732-252dd580-3f26-11eb-987b-57b0553c13bd.png)
> 
> Overall, it looks good as it matches Intel Coach results (+ a nice boost in final performance ~400 vs ~300 for DQN on Breakout) but as mentioned by @Miffyli that would nice to have at least one run with the original evaluation setting.

@araffin I tried to reproduce those results, but it did not work out. Which version of the games were used in your benchmark? https://www.gymlibrary.ml/environments/atari/ is documenting 3 versions (v0, v4, v5) and there a couple of different options for the v5 setting as well as different versions of the environment for v4. I tried "Breakout-v5" with default options as well as "BreakoutNoFrameskip-v4" however my learning curves do not look anything comparable to yours.
Code for reproduction (I know I could use `make_atari_env` directly, however for usage with my own monitoring wrappers not included in this example I need to build the environment manually in order to have proper access to the env without vec wrapper and stacking):

```
import gym
#from rl_utils.monitor import Monitor #my custom monitor
from sb3_contrib import QRDQN

from stable_baselines3.common.atari_wrappers import AtariWrapper
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.vec_env import VecFrameStack, DummyVecEnv

env = gym.make('BreakoutNoFrameskip-v4')
#gym.make('ALE/Breakout-v5', full_action_space=False, frameskip=1)
env = AtariWrapper(env)
env = DummyVecEnv([lambda: env])
env = VecFrameStack(env, n_stack=4)
#env = Monitor(env, params, "./results") # commented out, this is my custom monitor for recording learning curves
#env = make_atari_env(env, n_envs=1, seed=0)
model = QRDQN("CnnPolicy", env, verbose=1,
                             exploration_fraction=0.025,   # match QRDQN PR
                             seed=100)
model.learn(total_timesteps=10000000, log_interval=4)
```
My results looked like the following:

![Screenshot from 2022-07-04 10-11-12](https://user-images.githubusercontent.com/61889859/177115894-c41bec95-3691-405b-9670-35fcd99bc36b.png)
![Screenshot from 2022-07-04 10-12-24](https://user-images.githubusercontent.com/61889859/177115896-e106b3b1-da5b-44ad-ada4-464a808ebc8f.png)

Would really appreciate if you could give me hints on reproducing your baseline.

araffin commented:
Hello @hh0rva1h  ,

> Which version of the games were used in your benchmark? 

we always use the NoFrameskip-v4 (and gym 0.21).

> I tried to reproduce those results, but it did not work out. 

Please use the RL Zoo for that, we have instructions in the documentation: 
https://sb3-contrib.readthedocs.io/en/master/modules/qrdqn.html#how-to-replicate-the-results

Instructions are also on huggingface hub: https://huggingface.co/sb3/qrdqn-BreakoutNoFrameskip-v4

and detailed hyperparameters are here: https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/qrdqn.yml#L1


>  however for usage with my own monitoring wrappers not included in this example

I think your issue may come from the monitoring, please read https://github.com/DLR-RM/stable-baselines3/issues/181.

You should at least use `make_vec_env` and pass your custom wrappers.

mreitschuster commented:
> we always use the NoFrameskip-v4 (and gym 0.21).

any chance you have the reasoning in mind? I thought v4-noframeskip was to be avoided due to the "memorizing sequences" problem (even with up to 30 noops I suspect it just memorizes 31 sequences)

araffin commented:
>  any chance you have the reasoning in mind?

You can read more about why in the PR on updating gym: 
- https://github.com/DLR-RM/stable-baselines3/pull/572
- https://github.com/DLR-RM/stable-baselines3/pull/734

TL;DR: the new atari v5 were not yet benchmarked and as we have everything using `NoFrameskip-v4`, it is best to keep it like that so we can compare apples to apples.

> due to the "memorizing sequences" problem 

Do you have a reference for that?

mreitschuster commented:
[Revisiting the Arcade Learning Environment, Machado et al](https://arxiv.org/abs/1709.06009)
**page 7:** _The Arcade Learning Environment is also fully deter-
ministic – each game starts in the same state and outcomes are fully determined by the
state and the action. As such, it is possible to achieve high scores by learning an open-loop
policy, i.e., by simply memorizing a good action sequence, rather than learning to make
good decisions in a variety of game scenarios (Bellemare, Naddaf, Veness, & Bowling, 2015)._
**page 14** -> _Initial Noops ... The environment remains deterministic beyond the choice of starting state. Brute-like methods still perform well._

And to verify: train a model using NoFramskip-v4 and evaluate on v4 (stochasitcity via frameskip) or v5 (stochasticity via stickies) -> it will transfer poorly.
and then other way round: train of v4 or v5 and evaluate it on  NoFramskip-v4 -> good transfer.

I imagine a more sophisitcated approach would be to take randomly some i from 1-30 and use those Noop only in evaluation but not training. If model performs poorly on those frameskips on which it was not trained -> we know it just memorized.

mreitschuster commented:
@hh0rva1h 

> Do you have by any chance baselines curves at hand for the v4 or v5 scenarios? Would be really great to have more baselines apart from the NoFrameskip-v4. 

Not with qrdqn (yet). working on it. With PPO i have quite a few, but mostly with a non-standard wrapper configuration. But that would go off-topic from this PR. you can have a look at [my work on tuning & env selection](https://github.com/mreitschuster/RLbreakout/tree/master/learning/4_tuning). Short answer: without an aimbot I get to score of 180 and with aimbot to 220 (on 1e7 training steps).I havent found a PM functionality in github - if you want a discussion on that feel free to open an issue/discussion and add me there.

> Also, do you know by any chance how would one exactly match the atari setting as used in the the DQN and QR-DQN papers? ... would be closer to the paper setup than the baselines wrappers, would you agree?
Sorry, no & don't know. I was just testing models on breakout and wanted to check my results with others. Found this thread showing scores of 400 - but then realizing it is in the deterministic environment the excitement level dropped.

araffin commented:
Thanks @mreitschuster for the links!

> stochasitcity via frameskip

The RL Zoo models are using frameskip (via the `MaxAndSkip` wrapper) but the frame skipping is not stochastic, is that what you meant? (see https://github.com/DLR-RM/stable-baselines3/blob/a6f5049a99a4c21a6f0bcce458ca3306cef310e0/stable_baselines3/common/atari_wrappers.py#L223)

> train a model using NoFrameskip-v4 and evaluate on v4

you need to be careful when you do that, because as mentioned before, if you use atari wrapper, there is frame skipping.

> stochasticity via stickies

by sticky actions you mean randomly repeating actions?

mreitschuster commented:
@araffin 

> The RL Zoo models are using frameskip (via the MaxAndSkip wrapper) but the frame skipping is not stochastic, is that what you meant?
It is my understanding that we have two different frameskips available - stochastic frameskip (randomly skip frames) which is inbuilt into *-v4 (but deactivated when using *NoFrameskip-v4). it is also not active in v5 (which injects stochasticity diffferently).

and then there is the deterministic frameskip - found in [sb3 atariwrapper](https://stable-baselines3.readthedocs.io/en/master/_modules/stable_baselines3/common/atari_wrappers.html#MaxAndSkipEnv) as well as in the environment [*Deterministic-v4](https://github.com/openai/gym/issues/1280).

> you need to be careful when you do that, because as mentioned before, if you use atari wrapper, there is frame skipping.

yes you are right, but the wrapper provides determinstic frameskip (for speeding up the game), not the stochastic one (to train more transferable skills) 

>by sticky actions you mean randomly repeating actions?
yes. 

 i think if we want to deepen this (and i would be very happy to) it should be a seperate discussion.
